TFM: Allow tfm_ns_lock_dispatch() can dispatch in some interrupt-disable condition

ccli8 opened the pull request:
### Description

In TFM, to call secure function from non-secure world, we must go via `tfm_ns_lock_dispatch()` to guarantee thread-safe in secure world. Original implementation of `tfm_ns_lock_dispatch()` doesn't  allow this call in interrupt-disabled context for mutex acquire/release. This PR tries to relieve this limitation when:
1. Not in IRQ mode
1. IRQ masked
1. NS lock not owned by any thread

In this condition, NS lock can be seen as acquired by current thread and other threads cannot try to acquire it due to interrupt disabled, so the call to secure function would be safe.

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

kjbracey commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_ns_lock_rtx.c:
```diff
@@ -53,6 +56,20 @@ uint32_t tfm_ns_lock_dispatch(veneer_fn fn,
+     * 1. Not in IRQ mode
+     * 2. IRQ masked
+     * 3. NS lock not owned by any thread */
+    bool can_dispatch_with_irq_masked = (!IsIrqMode()) && IsIrqMasked() && (!ns_lock.lock_count);
```
In the case `!IsIrqMode() && IsIrqMasked() && ns_lock.lock_count`, you proceed to try to acquire the mutex, and crash because you can't acquire a mutex with interrupts disabled.

You should presumably return an error in that case.

Then, given the basic behaviour of this form has to be "try to do it if no-one else in mid-operation, else return error", I'm not sure why it's worth messing with atomics and counters outside the mutex. Why not just set a flag inside the mutex?

    osMutexAcquire();
    ns_lock = true;
    fn();
    ns_lock = false;
    os_MutexRelease();

That will work fine.

Although I'm not sure how useful it is to add a "interrupt disabled" mode that only works if another thread isn't touching security. Seems like a huge trap for developers - permitting secure calls with interrupts disabled that work 99.9% of the time, then start failing unexpectedly as the amount of secure-using threads increases.

ccli8 replied:
> Seems like a huge trap for developers - permitting secure calls with interrupts disabled that work 99.9% of the time, then start failing unexpectedly as the amount of secure-using threads increases.

Yes, it doesn't make sense. But during my NUMAKER_PFM_M2351 port, I find there's a code path which needs to call secure function in interrupt-disabled context:

<pre>
sleep_manager_sleep_auto > core_util_critical_section_enter > hal_sleep/hal_deepsleep > <b>CLK_Idle</b>/<b>CLK_PowerDown</b>
</pre>

Would you have solution to it?

kjbracey commented on components/TARGET_PSA/TARGET_TFM/COMPONENT_NSPE/interface/src/tfm_ns_lock_rtx.c:
```diff
@@ -53,6 +56,20 @@ uint32_t tfm_ns_lock_dispatch(veneer_fn fn,
+    }
+
+    if (core_util_atomic_incr_u16(&ns_lock.lock_count, 1) == 0) {
+        MBED_ERROR1(MBED_MAKE_ERROR(MBED_MODULE_UNKNOWN, MBED_ERROR_CODE_OVERFLOW), "NS lock count overflow (> 0xFFFF)", ns_lock.lock_count);
```
Given that the increment and decrement are in the same function, having all these checks seems overkill. Other lock/unlock in Mbed OS has the checks because they're separate functions and could be mismatched by the caller.

ccli8 committed:
TFM: Export tfm_ns_lock_get_init_state

When NS lock is not initialized yet, it means now is in pre-rtos stage and
non-secure world can call into secure world straight.

ccli8 committed:
TFM: Export tfm_ns_lock_get_lock_state

When NS lock is not locked and interrupt is also disabled, the current thread
can regard as having acquired NS lock and can call into secure world straight.
